### DISCUSSION TITLE
[Suggestions] System Options window, allow the use of the mouse wheel to change the setting of the options

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#4241] [Ideas]
When the mouse cursor is positioned over the graphic of the different options, allow the use of the mouse wheel to change the setting of the option, mouse wheel up to increase, and mouse wheel down to decrease.

In fheroes2:

![Graphic mouse wheel 01](https://user-images.githubusercontent.com/43583994/134171874-eb51d24d-5d08-4656-9ea0-f44964b2d906.png)

**Comment by a1exsh** [comment]
Had a hard time finding it.  Thought it was an issue, but it was a discussion. :)

Here's a PR that enables the wheel: https://github.com/ihhub/fheroes2/pull/4553

**Reply by LeHerosInconnu** [reply]
Hello @a1exsh,

> Had a hard time finding it. Thought it was an issue, but it was a discussion. :)

Yes, there are still a lot of issues and the discussions are building up too, and it's hard to get your head around it sometimes. :)

> Here's a PR that enables the wheel: #4553

I tried your PR, it works fine. :)

It is also worth noting that there are ten levels of sound adjustment besides "Off" in the original game ("On" corresponding to "Volume 10", and "On Volume 9" to "On Volume 1") and only nine in fheroes2.
The same goes for the Sound Effects.

Why this choice not to let the user cycle the values continuously with the mouse wheel?
It is less natural and less practical with the up and down stops.

It also seems a bit strange to be able to change some settings with the mouse wheel and not the others, the user has to do some "attention work" to use the mouse wheel with only some of the settings.

**Reply by a1exsh** [reply]
> It is less natural and less practical with the up and down stops.

I don't think so — check how system volume or screen brightness works with the mouse wheel on your OS, for example.

Wrapping over on click is left as a tribute to the OG.

> It also seems a bit strange to be able to change some settings with the mouse wheel and not the others

Yes and no: these are different enough to justify it, IMO. With sound volume or hero movement speed we have different levels of one thing (think volume bar) — less so with music type, interface or auto battle (think checkbox / two or tri-state).

**Reply by LeHerosInconnu** [reply]
@a1exsh.

> > It is less natural and less practical with the up and down stops.
> 
> I don't think so — check how system volume or screen brightness works with the mouse wheel on your OS, for example.
> 
> Wrapping over on click is left as a tribute to the OG.

As the user can use both ways, with the mouse click or with the mouse wheel, the operation is not similar in both cases (with the mouse click the cycle is continuous, there is no "back and forth" as with the mouse wheel), which is not good from the ergonomic point of view of using the interface.

> > It also seems a bit strange to be able to change some settings with the mouse wheel and not the others
> 
> Yes and no: these are different enough to justify it, IMO. With sound volume or hero movement speed we have different levels of one thing (think volume bar) — less so with music type, interface or auto battle (think checkbox / two or tri-state).

This is not so much because of the wide range of settings, but rather because of the ease and similarity of operation.
The user uses the mouse wheel for some settings and is forced to use the mouse button for others, which makes it less convenient.

**Reply by a1exsh** [reply]
> As the user can use both ways, with the mouse click or with the mouse wheel, the operation is not similar in both cases (with the mouse click the cycle is continuous, there is no "back and forth" as with the mouse wheel), which is not good from the ergonomic point of view of using the interface.

There's more than just that. Clicking only goes in the direction of increasing volume and wraps from maximum volume to off. I can imagine wrapping from zero to maximum volume is much less pleasant.

At the same time there is no way to turn the sound on or off completely without going through all the volume levels — either by clicking or using the wheel. What if we make clicking toggle the sound or music on/off and use the wheel for tuning the volume instead?

> The user uses the mouse wheel for some settings and is forced to use the mouse button for others, which makes it less convenient.

Sure, we don't have to restrict it to clicks only. I would not use it probably, but can't stop others from that. :)

**Reply by LeHerosInconnu** [reply]
Hello @a1exsh,

> > As the user can use both ways, with the mouse click or with the mouse wheel, the operation is not similar in both cases (with the mouse click the cycle is continuous, there is no "back and forth" as with the mouse wheel), which is not good from the ergonomic point of view of using the interface.
> 
> There's more than just that. Clicking only goes in the direction of increasing volume and wraps from maximum volume to off. I can imagine wrapping from zero to maximum volume is much less pleasant.

Yes, I understand the problem that can occur when the sound goes directly from zero to max volume. :)
To remedy this, the sound can be made to rise in volume in a smooth transition.

Like this:

![Son progression 01](https://user-images.githubusercontent.com/43583994/140608811-43b3a630-76f2-41e5-b7eb-acc1e6dc1f1a.png)

Instead of this:

![Son abrupte 02](https://user-images.githubusercontent.com/43583994/140608818-2a677e76-cf36-4f51-bb1c-679ce4efd304.png)

Another possibility would be to change the volume level adjustment sequence.

Like this:

Off
1
2
3
4
5
6
7
8
9
10
9
8
7
6
5
4
3
2
1
Off
1
2
3
4
5
6
7
8
9
10
9
8
7
6
5
4
3
2
1
etc.

Instead of this:

Off
10
9
8
7
6
5
4
3
2
1
Off
10
9
8
7
6
5
4
3
2
1
etc.

My preference goes for the second possibility. :)

> At the same time there is no way to turn the sound on or off completely without going through all the volume levels — either by clicking or using the wheel. What if we make clicking toggle the sound or music on/off and use the wheel for tuning the volume instead?

For the sake of similarity of operations all control buttons should work in the same way and everything should be possible with only the mouse button or only the mouse wheel.

> > The user uses the mouse wheel for some settings and is forced to use the mouse button for others, which makes it less convenient.
> 
> Sure, we don't have to restrict it to clicks only. I would not use it probably, but can't stop others from that. :)

Absolutely. :)

**Comment by a1exsh** [comment]
@ihhub what do you think about further tuning of the behavior discussed here — should we try to implement that as well?  I personally wouldn't use mouse wheel to change interface or music type, but that doesn't mean the possibility shouldn't be there.

**Reply by ihhub** [reply]
Before adding more mouse scroll logic I see that for Music and Sound volumes we don't have value 10 which is a bug as stated by @LeHerosInconnu . I think we should fix it first.
